### PR TITLE
Implement memstats command in the debugger

### PR DESCRIPTION
**Pull request by DanielBallaSZTE**
Implementation of memstats command in jerry-debugger, python and html client.
Shows the allocated bytes, byte code bytes, string bytes, object bytes, property bytes and heap size.

JerryScript-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/debugger/jerry-debugger.h [hunk 1]
```diff
@@ -217,6 +219,21 @@ typedef struct
 } jerry_debugger_receive_update_breakpoint_t;
 
 /**
+ * Outgoing message: send memory statistics
+ */
+ typedef struct
+ {
+  jerry_debugger_send_header_t header; /**< message header */
```
Comment (LaszloLango) [inline jerry-core/debugger/jerry-debugger.h]: wrong indentation

File: jerry-debugger/jerry-client-ws.py [hunk 2]
```diff
@@ -932,6 +940,22 @@ def main():
...
 
+        elif buffer_type == JERRY_DEBUGGER_MEMSTATS_RECEIVE:
+
+            memory_stats = struct.unpack(debugger.byte_order + debugger.idx_format
+                                        + debugger.idx_format + debugger.idx_format
+                                        + debugger.idx_format + debugger.idx_format
+                                        + debugger.idx_format,
+                                        data[3: 3 + 4 + 4 + 4 + 4 + 4 + 4])
+
+            print ("Allocated bytes: " + str(memory_stats[0]))
```
Comment (LaszloLango) [inline jerry-debugger/jerry-client-ws.py]: no space after `print`

File: jerry-core/debugger/jerry-debugger.c [hunk 3]
```diff
@@ -669,4 +678,56 @@ jerry_debugger_send_parse_function (uint32_t line, /**< line */
   return jerry_debugger_send (sizeof (jerry_debugger_send_parse_function_t));
 } /* jerry_debugger_send_parse_function */
 
+/**
+ * Send memory statistics to the debugger client.
+ */
+
+void
```
Comment (zherczeg) [inline jerry-core/debugger/jerry-debugger.c]: No newline before `void`.

File: jerry-core/debugger/jerry-debugger.c [hunk 4]
```diff
@@ -669,4 +678,56 @@ jerry_debugger_send_parse_function (uint32_t line, /**< line */
...
+  memset (memstats_p->object_bytes, 0, sizeof (uint32_t));
+  memset (memstats_p->property_bytes, 0, sizeof (uint32_t));
+
+#endif
+
+  jerry_debugger_send (sizeof (jerry_debugger_send_memstats_t));
+
+} /* jerry_debugger_send_memstats */
+
+
```
Comment (zherczeg) [inline jerry-core/debugger/jerry-debugger.c]: Only one newline here.

File: jerry-core/debugger/jerry-debugger.c [hunk 5]
```diff
@@ -669,4 +678,56 @@ jerry_debugger_send_parse_function (uint32_t line, /**< line */
...
+  memset (memstats_p->byte_code_bytes, 0, sizeof (uint32_t));
+  memset (memstats_p->string_bytes, 0, sizeof (uint32_t));
+  memset (memstats_p->object_bytes, 0, sizeof (uint32_t));
+  memset (memstats_p->property_bytes, 0, sizeof (uint32_t));
+
+#endif
+
+  jerry_debugger_send (sizeof (jerry_debugger_send_memstats_t));
+
+} /* jerry_debugger_send_memstats */
```
Comment (zherczeg) [inline jerry-core/debugger/jerry-debugger.c]: Remove the newline before }

File: jerry-core/debugger/jerry-debugger.h (hunk 1 above)
Comment (zherczeg) [inline jerry-core/debugger/jerry-debugger.h]: Two space indentation everywhere.

File: jerry-core/debugger/jerry-debugger.c [hunk 6]
```diff
@@ -669,4 +678,56 @@ jerry_debugger_send_parse_function (uint32_t line, /**< line */
...
+  uint32_t heap_size = (uint32_t) heap_stats->size;
+  memcpy (memstats_p->size, &heap_size, sizeof (uint32_t));
+
+#else /* if not, just put zeros */
+
+  memset (memstats_p->allocated_bytes, 0, sizeof (uint32_t));
+  memset (memstats_p->byte_code_bytes, 0, sizeof (uint32_t));
+  memset (memstats_p->string_bytes, 0, sizeof (uint32_t));
+  memset (memstats_p->object_bytes, 0, sizeof (uint32_t));
+  memset (memstats_p->property_bytes, 0, sizeof (uint32_t));
```
Comment (zherczeg) [inline jerry-core/debugger/jerry-debugger.c]: Heapsize should be cleared here as well. I am still not sure sending heapsize is useful.

File: jerry-debugger/jerry-client-ws.html [hunk 7]
```diff
@@ -806,6 +808,21 @@
         return;
       }
 
+
+      case JERRY_DEBUGGER_MEMSTATS_RECEIVE:
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.html]: Remove newline above.

File: jerry-debugger/jerry-client-ws.html [hunk 8]
```diff
@@ -806,6 +808,21 @@
...
+        var messagedata = decodeMessage("IIIIII", message, 1);
+
+        appendLog("Allocated bytes: " + messagedata[0]);
+        appendLog("Byte code bytes: " + messagedata[1]);
+        appendLog("String bytes: " + messagedata[2]);
+        appendLog("Object bytes: " + messagedata[3]);
+        appendLog("Property bytes: " + messagedata[4]);
+        appendLog("Heap size: " + messagedata[5]);
+
+        return;
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.html]: Remove newline above.

File: jerry-debugger/jerry-client-ws.py [hunk 9]
```diff
@@ -389,6 +391,12 @@ def do_exception(self, args):
 
         self.debugger.send_exception_config(enable)
 
+    def do_memstats(self, args):
+        """ Memory statistics """
+        self.exec_command(args, JERRY_DEBUGGER_MEMSTATS);
+        return
+
+
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: Remove two newlines above.

File: jerry-debugger/jerry-client-ws.py (hunk 2 above)
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: I think you used use the % format:

`"Allocated bytes: %d" % (str(memory_stats[0]))`

**Comment by DanielBallaSZTE** [comment]
Fixed those styling issues, and yes I also think it shouldn't be added afterall.

File: jerry-core/debugger/jerry-debugger.c [hunk 10]
```diff
@@ -669,4 +678,45 @@ jerry_debugger_send_parse_function (uint32_t line, /**< line */
   return jerry_debugger_send (sizeof (jerry_debugger_send_parse_function_t));
 } /* jerry_debugger_send_parse_function */
 
+/**
+ * Send memory statistics to the debugger client.
+ */
+
+void jerry_debugger_send_memstats (void)
```
Comment (zherczeg) [inline jerry-core/debugger/jerry-debugger.c]: No newline after `*/` command and the void must be in separate line

File: tests/debugger/do_help.expected [hunk 11]
```diff
@@ -4,8 +4,8 @@ Stopped at tests/debugger/do_help.js:15
 
 Documented commands (type help <topic>):
 ========================================
-b          bt        delete  eval       help  next        s   
-backtrace  c         dump    exception  list  pendingdel  src 
-break      continue  e       fbreak     n     quit        step
+b          bt        delete  eval       help      n           quit  step
```
Comment (zherczeg) [inline tests/debugger/do_help.expected]: Is python put the step in the same line????

Comment (DanielBallaSZTE) [inline tests/debugger/do_help.expected]: Yes, it does. `Documented commands (type help <topic>):
========================================
b          bt        delete  eval       help      n           quit  step
backtrace  c         dump    exception  list      next        s   
break      continue  e       fbreak     memstats  pendingdel  src 

(jerry-debugger) 
`

**Comment by DanielBallaSZTE** [comment]
As heap_size is not used anymore, I removed it.

File: jerry-debugger/jerry-client-ws.py [hunk 12]
```diff
@@ -932,6 +938,19 @@ def main():
 
             prompt.cmdloop()
 
+        elif buffer_type == JERRY_DEBUGGER_MEMSTATS_RECEIVE:
+
+            memory_stats = struct.unpack(debugger.byte_order + debugger.idx_format *5,
+                                        data[3: 3 + 4 *5])
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: Misaligned line (add a space).

File: jerry-debugger/jerry-client-ws.py [hunk 13]
```diff
@@ -389,6 +391,10 @@ def do_exception(self, args):
 
         self.debugger.send_exception_config(enable)
 
+    def do_memstats(self, args):
+        """ Memory statistics """
+        self.exec_command(args, JERRY_DEBUGGER_MEMSTATS);
+        return
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: We could add the short form (ms) here as well.

File: jerry-core/debugger/jerry-debugger.c [hunk 14]
```diff
@@ -347,6 +347,15 @@ jerry_debugger_process_message (uint8_t *recv_buffer_p, /**< pointer the the rec
       return true;
     }
 
+    case JERRY_DEBUGGER_MEMSTATS:
+    {
+      JERRY_DEBUGGER_CHECK_PACKET_SIZE (jerry_debugger_receive_type_t);
+
+      jerry_debugger_send_memstats ();
+
+      return true;
```
Comment (zherczeg) [inline jerry-core/debugger/jerry-debugger.c]: Remove newline before this return.

**Comment by LaszloLango** [comment]
Please fix the mentioned style issues, so we can merge the PR.

**Pull request merged**